src/dune_rules/expander0.ml

```diff
@@ -24,3 +24,26 @@ let as_in_build_dir ~what ~loc p =
           (Path.to_string_maybe_quoted p)
       ]
 ;;
+
+module type S = sig
+  type t
+
+  val project : t -> Dune_project.t
+  val eval_blang : t -> Blang.t -> bool Memo.t
+end
+
+open Memo.O
+
+type t = E : 'a Memo.t * (module S with type t = 'a) -> t
```

Comment:
Why are GADTs / witness necessary? Would it be possible to implement it with simpler types?